### PR TITLE
Added TSDB import with OpenMetrics and CSV file support.

### PR DESCRIPTION
**Pull request by bwplotka**
Based on https://github.com/prometheus/prometheus/pull/5887 Thanks for your work so far @dipack95, it helped a lot!

## Changes on top of @dipack95 and above PR:

* Addressed all reviews components
* Used subcommands for different formats
* Simplified block creation, no need to be such complex for the first iteration.
* Simplified OpenMetrics parsing from io.Reader, the previous implementation was over-allocating a lot, I think this might be simple start. @dipack95 please help to finish `openmetrics` package and tests :hugs: 
* Simplified and separate concerns. No need to have access to DB. Block
   * Block Writing is separated from parsing for ease of benchmarking and test. This will be also
needed by @JessicaGreben
* Added import support for different formats.
* Removed all tests - those had to be pulled over and adjusted ): 

I wanted to create a demo of CSV file import for KubeCon import, so went ahead and tried to help you @dipack95. Hope we can together finish this and add two formats soon :hugs: What do you think of using this (if you think it's ok, please give your feedback!). Feel free to propose changes (PRs) to this branch `tsdb-import` so we can have this merged together. Help wanted around tests and benchmarks!

cc @krasi-georgiev @JessicaGreben @dipack95 @brian-brazil 

### WIP (we can do some of those in separate PR)

* [ ] Adjust Docs
* [ ] Openmetrics package (parsing from io.Reader and tests)
* [ ] CSV parser support for different entries than series
* [ ] Tests
* [ ] Benchmark.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: docs/storage.md [hunk 1]
```diff
@@ -103,3 +103,57 @@ Note that on the read path, Prometheus only fetches raw series data for a set of
 ### Existing integrations
 
 To learn more about existing integrations with remote storage systems, see the [Integrations documentation](https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage).
+
+## Database management
```
Comment (bwplotka) [inline docs/storage.md:107]: To be adjusted

File: tsdb/cmd/tsdb/main.go [hunk 2]
```diff
@@ -52,29 +57,43 @@ func execute() (err error) {
...
+
+		dumpCmd     = cli.Command("dump", "dump samples from a TSDB")
+		dumpPath    = dumpCmd.Arg("db path", "database path (default is "+defaultDBPath+")").Default(defaultDBPath).String()
+		dumpMinTime = dumpCmd.Flag("min-time", "minimum timestamp to dump").Default(strconv.FormatInt(math.MinInt64, 10)).Int64()
+		dumpMaxTime = dumpCmd.Flag("max-time", "maximum timestamp to dump").Default(strconv.FormatInt(math.MaxInt64, 10)).Int64()
+
+		importCmd       = cli.Command("import", "[Experimental] import samples from input and produce TSDB block. Please refer to the storage docs for more details.")
+		importDbPath    = importCmd.Flag("output", "output directory for generated block").Default(".").String()
+		importFilePath  = importCmd.Flag("input-file", "disables reading from input and using file to import samples from. If empty input is required").String()
+		importBlockSize = importCmd.Flag("block-size", "The maximum block size. The actual block timestamps will be aligned with Prometheus time ranges.").Default("2h").Hidden().Duration()
```
Comment (brian-brazil) [inline tsdb/cmd/tsdb/main.go:85]: The help here doesn't seem right.

File: docs/storage.md [hunk 3]
```diff
@@ -103,3 +103,57 @@ Note that on the read path, Prometheus only fetches raw series data for a set of
...
+
+### Import old metrics from file
+
+The command-line tool `tsdb` contains an `import` command. The goal of this tool is not for a regular backfilling, don't forget prometheus is a pull based system by assumed choice.
+You can use it to restore another prometheus partial dump, or any metrics exported from any system. The only supported input format is [OpenMetrics](https://openmetrics.io/) which is the same as Prometheus exporters exposition.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/storage.md:112]: Yes this is something I did not touch yet, to be done :+1:

File: pkg/textparse/interface.go [hunk 4]
```diff
@@ -94,3 +95,45 @@ const (
 	MetricTypeStateset       = "stateset"
 	MetricTypeUnknown        = "unknown"
 )
+
+func (m *MetricType) ParseForOpenMetrics(mtyp string) error {
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply pkg/textparse/interface.go:99]: ack

File: tsdb/importer/openmetrics/openmetrics.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+	}
+	return 0, io.EOF
+}
+
+// Series returns the bytes of the series, the timestamp if set, and the value
+// of the current sample. Timestamps is milliseconds.
+func (p *Parser) Series() ([]byte, *int64, float64) {
+	b, ts, v := p.Parser.Series()
+	if ts != nil {
+		// OpenMetrics parser has time in nanoseconds. Convert to milliseconds.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply tsdb/importer/openmetrics/openmetrics.go:66]: ack, I copied what previous PR had (: Let's fix

File: tsdb/importer/csv/csv.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+)
+
+type HeaderType string
+
+const (
+	// TODO(bwplotka): Consider accepting different ones like `values` or `timestamps` and pairs of value - timestamp, similar to label name - value.
+	Value               HeaderType = "value"
+	TimestampMs         HeaderType = "timestamp_ms" // Milliseconds from seconds since 1970-01-01 00:00:00 UTC
+	MetricName          HeaderType = "metric_name"  // Value of "__name__" label.
+	LabelName           HeaderType = "label_name"   // If both labels and metric_name are specified labels are merged. This header can repeat, but label_value is expected as next column.
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply tsdb/importer/csv/csv.go:38]: makes sense

File: tsdb/importer/csv/csv.go [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+	LabelValue          HeaderType = "label_value"  // If both labels and metric_name are specified labels are merged. This header can repeat but label_name is expected to be in previous column.
+	Help                HeaderType = "help"
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply tsdb/importer/csv/csv.go:40]: Can agree. We don't even utilize this data yet, some maybe we can skip now.

File: tsdb/cmd/tsdb/main.go [hunk 8]
```diff
@@ -52,29 +57,43 @@ func execute() (err error) {
...
+
+		omImportCmd = importCmd.Command("openmetrics", "import samples from OpenMetrics input and produce TSDB block. Please refer to the storage docs for more details.")
+
+		csvImportCmd       = importCmd.Command("csv", "import samples from CSV input and produce TSDB block. Please refer to the storage docs for more details.")
+		csvImportDelimiter = csvImportCmd.Flag("delimiter", "CSV single character for fields delimiting").Default(",").String()
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply tsdb/cmd/tsdb/main.go:90]: I think it does not harm to allow configuring this (:

[the remark this answers is not in the document]
Comment (brian-brazil) [reply tsdb/cmd/tsdb/main.go:90]: I think there is harm, as it allows for more than one way to do things and thus confusion. Let's support one single CSV format - not N mildly different formats.

File: tsdb/importer/csv/csv.go [hunk 9]
```diff
@@ -0,0 +1,271 @@
+// Copyright 2020 The Prometheus Authors
```
Comment (JessicaGreben) [inline tsdb/importer/csv/csv.go:1]: could we move this to pkg/importer instead of tsdb?

**Comment by beorn7** [comment]
We (in the bug scrub) assume this has been superseded by #7675 and close this. Please re-open if we are wrong.

File: tsdb/importer/csv/csv.go (hunk 9 above)
Comment (bwplotka) [reply tsdb/importer/csv/csv.go:1]: :+1:

**Comment by bwplotka** [comment]
Reopening as those are two different things: Rule backfilling vs CSV backfilling.

**Comment by bwplotka** [comment]
Still happy to have some help / finish it (: Sorry for lag on this.

File: tsdb/importer/blocks/writer.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+	h, err := tsdb.NewHead(nil, logger, nil, durToMillis(9999*time.Hour), w.tmpDir, nil, tsdb.DefaultStripeSize, nil)
+	if err != nil {
+		return errors.Wrap(err, "tsdb.NewHead")
+	}
+
+	w.head = h
+	return w.head.Init(math.MinInt64)
+}
+
+// Appender is not thread-safe. Returned Appender is thread-save however.
```
[the remark this answers is not in the document]
Comment (beorn7) [reply tsdb/importer/blocks/writer.go:100]: And it's better to say "goroutine-safe". Or even better:

`// Appender must not be called concurrently. The returned Appender, however, can safely be used concurrently.`

**Review by beorn7 — COMMENTED** [review]
My plan is to let our intern pick this up in the week of 2020-09-07.

**Comment by codesome** [comment]
Closing this as I believe it is superseded by https://github.com/prometheus/prometheus/pull/7999

**Comment by omron93** [comment]
@codesome Why if anyone new is trying to solve the same issue all the previous PRs are immediately closed?
#7999 is a 'Draft' PR with most builds failing, this PR was working with most test jobs passing... so I think the current process of handling PRs isn't helping to solve the backfilling issue :-(

**Comment by beorn7** [comment]
The people working on #7999 and who have worked on this one are all collaborating closely. They have collectively decided it's best to use this PR as a PoC implementation but do the real one in a new PR, which is #7999.

**Pull request closed without merging**